tests/functional/api/groups/members_test.py

```diff
@@ -62,179 +62,135 @@ def test_it_returns_empty_list_if_no_members_in_group(self, app):
 
-        assert third_party_user in third_party_group.members
+    def test_it_errors_if_the_userid_is_invalid(self, do_request):
+        do_request(userid="INVALID_USERID", status=404)
 
-    def test_it_ignores_forwarded_user_header(
-        self, app, factories, db_session, auth_client
-    ):
-        third_party_user = factories.User(authority="thirdparty.com")
-        third_party_group = factories.Group(authority="thirdparty.com")
-        headers = self.authorization_header(auth_client)
-        user2 = factories.User(authority="thirdparty.com")
-        db_session.commit()
+    def test_it_errors_if_the_request_isnt_authenticated(self, do_request, headers):
+        del headers["Authorization"]
 
-        headers["X-Forwarded-User"] = third_party_user.userid
+        do_request(status=404)
 
-        res = app.post_json(
-            "/api/groups/{pubid}/members/{userid}".format(
-                pubid=third_party_group.pubid, userid=third_party_user.userid
-            ),
-            headers=headers,
-        )
+    def test_it_errors_if_the_request_has_token_authentication(
+        self, factories, user, headers, do_request
+    ):
+        token = factories.DeveloperToken(user=user)
+        headers.update(token_authorization_header(token))
 
-        assert third_party_user in third_party_group.members
-        assert user2 not in third_party_group.members
-        assert res.status_code == 204
+        do_request(status=404)
 
-    def test_it_is_idempotent(self, app, auth_client, db_session, factories):
-        third_party_user = factories.User(authority="thirdparty.com")
-        third_party_group = factories.Group(authority="thirdparty.com")
-        db_session.commit()
+    def test_it_errors_if_the_groups_authority_doesnt_match(
+        self, factories, do_request
+    ):
+        group = factories.Group(authority="other")
 
-        app.post_json(
-            "/api/groups/{pubid}/members/{userid}".format(
-                pubid=third_party_group.pubid, userid=third_party_user.userid
-            ),
-            headers=self.authorization_header(auth_client),
-        )
+        do_request(pubid=group.pubid, status=404)
 
-        res = app.post_json(
-            "/api/groups/{pubid}/members/{userid}".format(
-                pubid=third_party_group.pubid, userid=third_party_user.userid
-            ),
-            headers=self.authorization_header(auth_client),
-        )
+    def test_it_errors_if_the_users_authority_doesnt_match(self, factories, do_request):
+        user = factories.User(authority="other")
 
-        assert third_party_user in third_party_group.members
-        assert res.status_code == 204
+        do_request(userid=user.userid, status=404)
 
-    def test_it_returns_404_if_authority_mismatch_on_user(
-        self, app, factories, auth_client
+    def test_it_errors_if_the_groups_and_users_authorities_both_dont_match(
```

Comment:
This is a dangerous case that was missing from the previous tests: the view code contains [a test that the user's and group's authorities match](https://github.com/hypothesis/h/blob/1870fa8ec3386df8ae4b82820dcdbf5e517caa83/h/views/api/groups.py#L204-L205). But what if the user and group's authorities are the same but they _both_ differ from that of the authclient? This could allow authclients to add members to groups outside of the authclient's authority! It turns out there's no bug here because the `ADD_MEMBER` permission [requires the group's authority to match the authclient's](https://github.com/hypothesis/h/blob/1870fa8ec3386df8ae4b82820dcdbf5e517caa83/h/security/permission_map.py#L57). But still I'm glad to have a functest for this case.